File: subsets.js
//Objective is, given a set of distinct integers, return all possible subsets.

let nums = [1,2,3]


//O(N*2^N) solution that has no if condition when backtracking

let temp = []
let index = 0
let result = []

backtrack(temp, index)

function backtrack(temp, index) {
    result.push([...temp])

    for (let i = index; i < nums.length; i++) {
        backtrack([...temp, nums[i]], i + 1)
    }
}

return result

File: subsets.java
//Java Solution

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        LinkedList<Integer> temp = new LinkedList<>();
        int index = 0;
        
        backtrack(temp, result, index, nums);
        return result;
    }
    
    public void backtrack(LinkedList<Integer> temp, List<List<Integer>> result, int index, int[] nums) {
        result.add(new ArrayList<>(temp));
        
        for (int i = index; i < nums.length; i++) {
            temp.add(nums[i]);
            backtrack(temp, result, i + 1, nums);
            temp.removeLast();
        }
    }
}